dulwich/client.py

```diff
@@ -502,7 +502,10 @@ def fetch(self, path, target, determine_wants=None, progress=None, depth=None):
 
         """
         if determine_wants is None:
-            determine_wants = target.object_store.determine_wants_all
+            if depth is not None and depth > 1:
+                determine_wants = target.object_store.determine_wants_force
```

Comment:
determine_wants is meant to be used by the caller to specify what refs they're interested in. What's the point of asking them, if we're going to fetch the ref anyway?

ideally, this should be checking what depth we're currently at and whether that's less than what was requested.

Reply:
My main issue here was that `porcelain.fetch` does not take a `determine_wants` callback at all, and only uses the default  `object_store.determine_wants_all`. If it would be preferred to update `porcelain.fetch` to accept the callback (and pass it through to `client.fetch` I can go ahead and make that change

Reply:
Perhaps an option would be to add an optional depth argument to determine_wants, which gets specified when there is a depth?

That way, target.object_store.determine_wants() can return refs depending on how deep the history of target.object_store is.

Reply:
(to start off with, you could use the current logic - always fetch if depth is > 1, but that way we can make it more correct later by actually checking the current depth)

Reply:
Updated this so that depth is now passed into `determine_wants`. For the default callback, we will always fetch everything if a complete unshallow is requested via `depth=DEPTH_INFINITE`, otherwise we will check the current available depth for a ref